### ISSUE TITLE
Support for dynamic code actions

### ISSUE THREAD

**Issue by Gelio** [mattn/efm-langserver#114]
Hey,

I've been experimenting with neovim's LSP which got me into using this language server (great work btw) for https://github.com/mantoni/eslint_d.js. It's been working nicely so far, but the one thing I've been missing is the ability to invoke a single quick fix provided by ESLint. It has also been mentioned in https://github.com/mattn/efm-langserver/issues/86#issuecomment-766716178 that the author could not get code actions to work.

I am aware that one can use the `formatCommand` to apply all fixes this way, but this:
1. Does not allow applying fixes selectively (only a specific fix)
2. Does not inform how many problems will be fixed. It's more of a _shoot fire aim_ approach
3. I could not get it to work alongside https://github.com/fsouza/prettierd - I could either get prettier formatting, or ESLint autofix.

I've been thinking of adding support for code actions to this language server and to later get it working in eslint_d. So far, I have noticed that there is some support for code actions (the `Command` struct), but I believe it only supports a known static set of commands provided in the language configuration.

Would you be willing to add support for getting a list of code actions from a given command? I know this is a broad topic, and I'm willing to help

**Comment by vatosarmat** [comment]
There is already [dedicated language server](https://github.com/hrsh7th/vscode-langservers-extracted) for `eslint` and it is supported in `neovim` [almost out-of-the-box](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/eslint.lua). So this issue is no longer applicable and can be closed

**Comment by Gelio** [comment]
True - either use https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#eslint or https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md#eslint-2

**Comment by Isti115** [comment]
Sorry to bump a closed issue, but with [null-ls being archived soon](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/1621), I think that this might be worth a second thought.

**Comment by mrjones2014** [comment]
Agreed. I'll really miss the code actions for shellcheck and statix.

**Comment by Gelio** [comment]
I agree, efm-langserver seems like a viable successor to null-ls. I will leave it up to @mattn to decide whether to reopen this issue since I don't know whether he intends to add dynamic actions to this project.

In general, I am not sure how easy it would be to get a list of code actions from a tool like ESLint or shellcheck. Perhaps the format of code actions would be configured for a given CLI tool and then the exact commands would be generated based on that CLI's output. That's just a rough idea.

**Comment by llllvvuu** [comment]
So EFM is a config-based system and I feel like generating code actions is more of a code-based task. Possibly it would be best to write each code actions adapter as its own LSP (like what has been done for ESLint), or if there's value in having a central bus for everything (to handle debounce/etc), to make it in TypeScript (so that adapters can be written in TypeScript, like how null-ls code actions are written in Lua).

The downside of having a central bus is that the adapters that people write can't be used by themselves, i.e. they are dependent on the bus (i.e. efm/null-ls). And also there are cases where it doesn't align perfectly with LSP's model which reports capabilities per language server rather than per language I believe. So one LSP per adapter/integration makes much sense. Microsoft has a [pretty good guide for TypeScript](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide) (skip the VSCode language client part), and there's also [`tower-lsp-boilerplate`](https://github.com/IWANABETHATGUY/tower-lsp-boilerplate/blob/main/src/main.rs) for Rust stans.
